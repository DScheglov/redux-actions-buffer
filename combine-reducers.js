module.exports = require('./lib/combine-reducers');
