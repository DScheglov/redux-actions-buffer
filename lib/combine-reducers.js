'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.actionsBuffer = exports.combineReducers = undefined;

var _redux = require('redux');

var _ = require('./');

var combineReducers = exports.combineReducers = (0, _redux.compose)(_.enableBatching, _redux.combineReducers);

exports.actionsBuffer = _.actionsBuffer;