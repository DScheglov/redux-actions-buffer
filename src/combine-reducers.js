import { compose, combineReducers as reduxCombineReducers } from 'redux';
import { enableBatching, actionsBuffer } from './';

export const combineReducers = compose(
  enableBatching,
  reduxCombineReducers
);

export { actionsBuffer };