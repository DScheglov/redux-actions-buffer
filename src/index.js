const START_BUFFER = 'RDX-BTCH-BFFR::START';
const FLUSH_BUFFER = 'RDX-BTCH-BFFR::FLUSH';
const DESTROY_BUFFER = 'RDX-BTCH-BFFR::DESTROY';
const PROCESS_BATCH = 'RDX-BTCH-BFFR::BATCH';

export const startBuffering = { type: START_BUFFER };
export const flushBuffer = { type: FLUSH_BUFFER };
export const destroyBuffer = { type: DESTROY_BUFFER };

const argsArr = a => a.length && Array.isArray(a[0]) ? a[0] : a;

export const batch = (...actions) => ({
  type: PROCESS_BATCH,
  actions: argsArr(actions),
});


export const enableBatching = reducer => (state, action) => (
  action.type === PROCESS_BATCH
    ? action.actions.reduce(reducer, state)
    : reducer(state, action)
);

const bufferDestroy = bufferContainer => res => (
  (bufferContainer.buffer = null),
  res
);

const bufferFlush = bufferContainer => next => (
  bufferContainer.buffer != null &&
  bufferContainer.buffer.length > 0 &&
  bufferDestroy(bufferContainer)(
    next(batch(bufferContainer.buffer))
  )
);

const createBuffer = bufferContainer => next => (
  bufferFlush(bufferContainer)(next),
  bufferContainer.buffer = []
);

const processAction = bufferContainer => (next, action) => (
  bufferContainer.buffer ? bufferContainer.buffer.push(action) : next(action)
);

const bufferManager = bufferContainer => ({
  create: createBuffer(bufferContainer),
  flush: bufferFlush(bufferContainer),
  destroy: bufferDestroy(bufferContainer),
  process: processAction(bufferContainer),
});

export const actionsBuffer = store => {
  const buffer = bufferManager({});
  return next => action => (
    action.type === START_BUFFER ? buffer.create(next) :
    action.type === FLUSH_BUFFER ? buffer.flush(next) :
    action.type === DESTROY_BUFFER ? buffer.destroy(true) :
    buffer.process(next, action)
  );
}

export const arrayToBatch = _ => next => action => (
  Array.isArray(action) ? next(batch(action)) : next(action)
);