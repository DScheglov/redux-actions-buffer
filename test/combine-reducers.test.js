import { createStore, applyMiddleware, compose } from 'redux';

import { combineReducers, actionsBuffer } from '../src/combine-reducers';
import { mainSuite } from './index.test';

const createAppStoreImpl = (reducers, initialState = {}) => createStore(
  combineReducers(reducers),
  initialState,
  applyMiddleware(actionsBuffer)
);

mainSuite('Implicitly applied High Ordered Reducer', createAppStoreImpl);