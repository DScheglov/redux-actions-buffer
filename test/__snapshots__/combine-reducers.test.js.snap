// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Explicitly applied High Ordered Reducer created store should bufferize actions and then process the batch 1`] = `
Object {
  "lastActionType": "@@redux/INIT",
}
`;

exports[`Explicitly applied High Ordered Reducer creating a store 1`] = `
Object {
  "allActions": Array [
    Object {
      "type": "@@redux/INIT",
    },
  ],
}
`;

exports[`Implicitly applied High Ordered Reducer created store should bufferize actions and then process the batch 1`] = `
Object {
  "lastActionType": "@@redux/INIT",
}
`;

exports[`Implicitly applied High Ordered Reducer creating a store 1`] = `
Object {
  "allActions": Array [
    Object {
      "type": "@@redux/INIT",
    },
  ],
}
`;
